src/libopensc/card-openpgp.c

```diff
@@ -1083,20 +1090,30 @@ static int
 pgp_read_blob(sc_card_t *card, pgp_blob_t *blob)
 {
 	struct pgp_priv_data *priv = DRVDATA(card);
+	static u8 *buffer = NULL;
+	static size_t buf_size = 0;
 
 	if (blob->data != NULL)
 		return SC_SUCCESS;
 	if (blob->info == NULL)
 		return blob->status;
 
+	/* calculate buffer size required and allocate static buffer */
+	if (buffer == NULL) {
```

Comment:
how can this *not* be NULL?

Reply:
Because of the properties of variables declared using the storage class `static`!
In C, static variables survive outside the scope in which they are declared.
When re-entering the scope, the variables are in the same state they were in when the scope was previously left.
The assignment of a value in the declaration of a `static` variable is performed only once, i.e. when entering the scope the first time.

What the code does, is:
1) it declares the static variables, giving them their initial state
2) if the variables are still in their initial state, it calculates the size of the buffer and allocates the memory for it.
  If the variables are not in their initial state any more (i.e. buffer already allocated), then nothing is done.

Net result: buffer is allocated with the required size only once during the lifetime of the process.